src/middleware/jira-jwt-middleware.ts

```diff
@@ -1,41 +1,104 @@
-import { Installation } from "models/installation";
+import { decodeSymmetric, getAlgorithm } from "atlassian-jwt";
 import { NextFunction, Request, Response } from "express";
-import { sendError, TokenType, verifySymmetricJwtTokenMiddleware } from "../jira/util/jwt";
+import { extractJwtFromRequest, TokenType, verifySymmetricJwtAndSetResponseCodeOnError } from "~/src/jira/util/jwt";
+import { Installation } from "~/src/models/installation";
 
-export const verifyJiraJwtMiddleware = (tokenType: TokenType) => async (
-	req: Request,
-	res: Response,
-	next: NextFunction
-): Promise<void> => {
-	const { jiraHost } = res.locals;
+export const jiraHostMiddleware = async (req: Request, res: Response, next: NextFunction) => {
 
-	if (!jiraHost) {
-		sendError(res, 401, "Unauthorised");
+	const token = extractJwtFromRequest(req);
+
+	if (!token) {
+		if (req.session.jiraHost) {
+			res.locals.jiraHost = req.session.jiraHost;
+			next();
+			return;
+		}
+		next();
 		return;
 	}
 
-	const installation = await Installation.getForHost(jiraHost);
-
+	const installation = await getInstallationAndTokenVerification(token, req, res);
 	if (!installation) {
-		return next(new Error("Not Found"));
+		return;
 	}
-	// TODO: Probably not the best place to set things globally
+
 	res.locals.installation = installation;
+	res.locals.jiraHost = installation.jiraHost;
+	res.locals.jwtVerified = true;
+
+	if (req.cookies.jwt) {
+		req.session.jiraHost = installation.jiraHost;
```

Comment:
why this line `req.session.jiraHost = installation.jiraHost;`?

Reply:
This is a case when application accessed directly and jwt token migrated as a cookie. In that case jirahost to be kept in session and remove jwt cookie.